File: ios/Runner/FlutterSceneView/FlutterSceneViewModel.swift
//
//  FlutterSceneViewModel.swift
//  Runner
//
//  Created by Wayne Rumble on 05/10/2020.
//

import Foundation
import KMShared
import RxSwift

protocol FlutterSceneViewModelType {
    var colorBehaviourSubject: BehaviorSubject<String> { get }
}

class FlutterSceneViewModel: FlutterSceneViewModelType {
    let colorBehaviourSubject = BehaviorSubject(value: "#FFFFFF")
    private let database: Database

    required init(database: Database) {
        self.database = database

        listenToBackgroundColorFlow()
    }

    private func listenToBackgroundColorFlow() {
        database.backgroundColorFlow.watch { color in
            guard let colorHex = color?.hex else {
                return
            }
            self.colorBehaviourSubject.onNext(colorHex)
        }
    }
}

File: android/app/src/main/kotlin/com/rumblewayne/BIMMultiPlatform/FlutterSceneViewModel.kt
package com.rumblewayne.BIMMultiPlatform

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.rumblewayne.bimmultiplatform.db.cache.Database
import kotlinx.coroutines.flow.MutableStateFlow

class FlutterSceneViewModel(private val database: Database): ViewModel() {
    var colorFlow = MutableStateFlow<String>("#FFFFFF")

    init {
        viewModelScope
        listenToBackgroundColorFlow()
    }

    private fun listenToBackgroundColorFlow() {
        database.backgroundColorFlow.watch {
            colorFlow.value = it.hex
        }
    }
}